From e6e1849876ed62c1d9c96ba4e1b42f26dafe23ee Mon Sep 17 00:00:00 2001
From: Nadav Haklai <nadavh@marvell.com>
Date: Mon, 21 Sep 2015 10:56:31 +0300
Subject: [PATCH 0029/1345] ARM: mvebu: Disable CPU freq on Armada 38x for the
 multi CPU flavors

commit  fb31c2ba1f2c598a17b28d19bc81c3f65ef7304f from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

On multi CPU Armada 38x SoCs, under heavy I/O load, the system hangs when
cpufreq is enabled.
Waiting for a solution to this issue, this patch disables the cpufreq
support for those SoC.

The single CPU flavor of Armada 38x, Armada 380 does not suffer from the
cpufreq related system hang and therefore cpufreq can be enabled on this
platform.

Change-Id: Iae35574c2d538c01e22206724cf708691ba717b7
Signed-off-by: Nadav Haklai <nadavh@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/23688
Tested-by: Star_Automation <star@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/27314
Tested-by: Lior Amsalem <alior@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/cpufreq/mvebu-cpufreq.c |    6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/cpufreq/mvebu-cpufreq.c b/drivers/cpufreq/mvebu-cpufreq.c
index c6db532..292e2ba 100644
--- a/drivers/cpufreq/mvebu-cpufreq.c
+++ b/drivers/cpufreq/mvebu-cpufreq.c
@@ -37,6 +37,12 @@ static int __init mvebu_pmsu_cpufreq_init(void)
 	    !of_machine_is_compatible("marvell,armada380"))
 		return 0;
 
+	if (of_machine_is_compatible("marvell,armada380") &&
+	    (num_possible_cpus() > 1)) {
+		pr_warn("CPU freq is currently broken on Armada 38x: disabling");
+		return 0;
+	}
+
 	/*
 	 * In order to have proper cpufreq handling, we need to ensure
 	 * that the Device Tree description of the CPU clock includes
-- 
1.7.9.5

